test/support/sycl_alloc_utils.h

```diff
@@ -0,0 +1,129 @@
+// -*- C++ -*-
+//===-- sycl_alloc_utils.h ------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __TEST_SYCL_ALLOC_UTILS_H
+#define __TEST_SYCL_ALLOC_UTILS_H
+
+#ifdef TEST_DPCPP_BACKEND_PRESENT
+
+#include "oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h"
+
+namespace TestUtils
+{
+
+// RAII service class to allocate shared/device memory (USM)
+// Usage model"
+// 1. allocate USM memory and copying data to USM:
+//    sycl_usm_alloc<alloc_type, _ValueType> alloc(queue, first, count); 
+// or 
+//    sycl_usm_alloc<alloc_type, _ValueType> alloc(queue, std::begin(data), std::end(data));
+// or just allocate USM memory"
+//    sycl_usm_alloc<alloc_type, _ValueType> alloc(queue, count); 
+// 2. get a USM pointer by sycl_usm_alloc::get_data() and passed one into a parallel algorithm with dpc++ policy.
+// 3. Retrieve data back (in case of device allocation type) to the host for further checking result.
+//    alloc.retrieve_data(dest_host);
+template<sycl::usm::alloc _alloc_type, typename _ValueType>

```

Comment:
Overall this approach looks like overkill for me. I completely agree with idea of tag dispatching for different allocation and I am supportive to RAII approach but I think that real allocation function with that dispatching and then `unique_ptr` with custom deleter would be sufficient.

Reply:
I am implemented this variant before. The test code becomes quite complex and does not look pretty.